Allow the node channel to connect using SSL.

hyowon opened the pull request:
This CL adds "--secure-connection" switch to allow the node channel to connect securely by using BoringSSL. Temporarily a self-signed certificate file is being used.

hyowon committed:
Allow the node channel to connect using SSL.

This CL adds "--secure-connection" switch to allow the node
channel to connect securely by using BoringSSL.
Temporarily a self-signed certificate file is being used.

nke94 commented on mojo/public/cpp/platform/secure_socket_utils_posix.cc:
```diff
@@ -0,0 +1,111 @@
+class SSLClientContext {
+ public:
+  static SSLClientContext* GetInstance() {
+    return base::Singleton<SSLClientContext,
```
Could you explain the reason why it should be a singleton?
Based on this implementation, when two client socket is created and want to make secure connection, it will share same SSL context.
Is it normal way of usage open_SSL library?

hyowon replied:
We do not need different certificates and different configuration contexts. So whenever we need a secure connection, we can provide an individual SSL instance from a global singleton SSL context.

nke94 commented on mojo/core/channel_posix.cc:
```diff
@@ -509,6 +526,10 @@ class ChannelPosix : public Channel,
       ssize_t read_result =
+#if defined(CASTANETS)
+          secure_connection_ ?
+          SecureSocketRecvmsg(ssl_.get(), buffer, buffer_capacity) :
```
When secure_connection true.. is vector<base::ScopedFD> incoming_fds handled properly?
If handling incomming_fds_ is not meaningful for secure connection why don't you exclude below [for (auto& fd : incoming_fds)] block in case of secure_connection

hyowon replied:
As you say, incoming_fds is not used and empty when secure_connection. I just wanted to make the code diff by defined(CASTANETS) as small and simple as possible.
Doing [for (auto& fd: incoming_fds)] don't burden at all (because it should be empty), so why don't you maintain code simplicity?

hyowon replied:
I've modified what you mentioned.

hyowon committed:
Generate a self-signed certificate file.

Each time Castanets runs with the secure-connection switch, it
generates and uses a self-signed certificate file.
